glean-core/src/metrics/dual_labeled_counter.rs

```diff
@@ -0,0 +1,375 @@
+//                 .map_or(0, |v| unsafe { ops.malloc_enclosing_size_of(v) })
+//         } else {
+//             map.capacity()
+//                 * (mem::size_of::<String>() + mem::size_of::<T>() + mem::size_of::<usize>())
+//         };
+
+//         let mut map_size = shallow_size;
+//         for (k, v) in map.iter() {
+//             map_size += k.size_of(ops);
+//             map_size += v.size_of(ops);
+//         }
+
+//         self.labels.size_of(ops) + self.submetric.size_of(ops) + map_size
+//     }
+// }
+
+impl DualLabeledCounterMetric {
+    /// Creates a new dual labeled counter from the given metric instance and optional list of labels.
+    pub fn new(
+        meta: CommonMetricData,
+        keys: Option<Vec<Cow<'static, str>>>,
+        catgories: Option<Vec<Cow<'static, str>>>,
+    ) -> DualLabeledCounterMetric {
+        let submetric = CounterMetric::new(meta);
+        DualLabeledCounterMetric::new_inner(submetric, keys, catgories)
+    }
+
+    fn new_inner(
+        counter: CounterMetric,
+        keys: Option<Vec<Cow<'static, str>>>,
+        categories: Option<Vec<Cow<'static, str>>>,
+    ) -> DualLabeledCounterMetric {
+        let dual_label_map = Default::default();
+        DualLabeledCounterMetric {
+            keys,
+            categories,
+            counter,
+            dual_label_map,
+        }
+    }
+
+    /// Creates a new metric with a specific label.
+    ///
+    /// This is used for dynamic labels where we have to actually validate and correct the
+    /// label later when we have a Glean object.
+    fn new_counter_metric(&self, key: &str, category: &str) -> CounterMetric {
+        // Determine which type of dynamic label it is
+        if self.keys.is_none() && self.categories.is_none() {
+            self.counter
+                .with_dynamic_label(DynamicLabelType::KeyAndCategory(
+                    make_label_from_key_and_category(key, category),
+                ))
+        } else if self.keys.is_none() {
+            let static_category = self.static_category(category);
+            self.counter.with_dynamic_label(DynamicLabelType::KeyOnly(
+                make_label_from_key_and_category(key, static_category),
+            ))
+        } else if self.categories.is_none() {
+            let static_key = self.static_key(key);
+            self.counter
+                .with_dynamic_label(DynamicLabelType::CategoryOnly(
+                    make_label_from_key_and_category(static_key, category),
+                ))
+        } else {
+            // Both labels are static and can be validated now
+            let static_key = self.static_key(key);
+            let static_category = self.static_category(category);
+            let name = combine_base_identifier_and_labels(
+                self.counter.meta().inner.name.as_str(),
+                static_key,
+                static_category,
+            );
+            self.counter.with_name(name)
+        }
```

Comment:
I find if/elseif/else chains a bit hard to read.
What do you think about using a match here?

```rust
        match (&self.keys, &self.categories) {
            (None, None) => /* ... */
            (None, _) => /* ... */
            (_, None) => /* ... */
            _  => /* ... */
        }
```